File: LXExtension/UIKit+LXExtension/CALayer+LXExtension.m
//
//  CALayer+LXExtension.m
//
//  Created by 从今以后 on 15/10/5.
//  Copyright © 2015年 apple. All rights reserved.
//

#import "CALayer+LXExtension.h"

NS_ASSUME_NONNULL_BEGIN

@interface _LXAnimationDelegate : NSObject
@end
@implementation _LXAnimationDelegate {
    void (^_completion)(BOOL finished);
}

- (instancetype)initWithCompletion:(void (^)(BOOL finished))completion
{
    self = [super init];
    if (self) {
        _completion = completion;
    }
    return self;
}

- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag
{
    _completion(flag);
}

@end

@implementation CALayer (LXExtension)

#pragma mark - Bounds|Frame -

#pragma mark size

- (void)setLx_size:(CGSize)lx_size
{
    CGRect frame = self.frame;
    frame.size   = lx_size;
    self.frame   = frame;
}

- (CGSize)lx_size
{
    return self.frame.size;
}

- (void)setLx_width:(CGFloat)lx_width
{
    CGRect frame     = self.frame;
    frame.size.width = lx_width;
    self.frame       = frame;
}

- (CGFloat)lx_width
{
    return CGRectGetWidth(self.frame);
}

- (void)setLx_height:(CGFloat)lx_height
{
    CGRect frame      = self.frame;
    frame.size.height = lx_height;
    self.frame        = frame;
}

- (CGFloat)lx_height
{
    return CGRectGetHeight(self.frame);
}

#pragma mark origin

- (void)setLx_origin:(CGPoint)lx_origin
{
    CGRect frame = self.frame;
    frame.origin = lx_origin;
    self.frame   = frame;
}

- (CGPoint)lx_origin
{
    return self.frame.origin;
}

- (void)setLx_originX:(CGFloat)lx_originX
{
    CGRect frame   = self.frame;
    frame.origin.x = lx_originX;
    self.frame     = frame;
}

- (CGFloat)lx_originX
{
    return CGRectGetMinX(self.frame);
}

- (void)setLx_originY:(CGFloat)lx_originY
{
    CGRect frame   = self.frame;
    frame.origin.y = lx_originY;
    self.frame     = frame;
}

- (CGFloat)lx_originY
{
    return CGRectGetMinY(self.frame);
}

#pragma mark position

- (void)setLx_positionX:(CGFloat)lx_positionX
{
    CGPoint position = self.position;
    position.x       = lx_positionX;
    self.position    = position;
}

- (CGFloat)lx_positionX
{
    return self.position.x;
}

- (void)setLx_positionY:(CGFloat)lx_positionY
{
    CGPoint position = self.position;
    position.y       = lx_positionY;
    self.position    = position;
}

- (CGFloat)lx_positionY
{
    return self.position.y;
}

#pragma mark - 动画 -

- (void)lx_addAnimationWithDuration:(NSTimeInterval)duration
                            keyPath:(nullable NSString *)keyPath
                         animations:(void (^)(CABasicAnimation *animation))animations
{
    [self lx_addAnimationWithDuration:duration
                                delay:0
                              keyPath:keyPath
                                  key:nil
                           animations:animations
                           completion:nil];
}

- (void)lx_addAnimationWithDuration:(NSTimeInterval)duration
                            keyPath:(nullable NSString *)keyPath
                         animations:(void (^)(CABasicAnimation *animation))animations
                         completion:(nullable void (^)(BOOL finished))completion
{
    [self lx_addAnimationWithDuration:duration
                                delay:0
                              keyPath:keyPath
                                  key:nil
                           animations:animations
                           completion:completion];
}

- (void)lx_addAnimationWithDuration:(NSTimeInterval)duration
                              delay:(NSTimeInterval)delay
                            keyPath:(nullable NSString *)keyPath
                         animations:(void (^)(CABasicAnimation *animation))animations
                         completion:(nullable void (^)(BOOL finished))completion
{
    [self lx_addAnimationWithDuration:duration
                                delay:delay
                              keyPath:keyPath
                                  key:nil
                           animations:animations
                           completion:completion];
}

- (void)lx_addAnimationWithDuration:(NSTimeInterval)duration
                              delay:(NSTimeInterval)delay
                            keyPath:(nullable NSString *)keyPath
                                key:(nullable NSString *)key
                         animations:(void (^)(CABasicAnimation *animation))animations
                         completion:(nullable void (^)(BOOL finished))completion
{
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:keyPath];
    animation.duration  = duration;
    animation.beginTime = CACurrentMediaTime() + delay;
    animation.fillMode  = kCAFillModeBackwards;
    animations(animation);
    if (completion) {
        animation.delegate = [[_LXAnimationDelegate alloc] initWithCompletion:completion];
    }
    [self addAnimation:animation forKey:key];
}

@end

NS_ASSUME_NONNULL_END

File: LXUtilities/Extensions/UIKitExtensions/Swift/CALayer+LXExtension.swift
//
//  CALayer+LXExtension.swift
//
//  Created by 从今以后 on 16/1/10.
//  Copyright © 2016年 从今以后. All rights reserved.
//

import UIKit

extension Swifty where Base: CALayer {

	private class AnimationDelegate: NSObject, CAAnimationDelegate {
		var completion: ((Bool) -> Void)?
		init(completion: @escaping (Bool) -> Void) {
			self.completion = completion
		}
		func animationDidStart(_ anim: CAAnimation) {}
		func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
			completion?(flag)
			completion = nil
		}
	}

	/**
	添加动画。

	- parameter anim:       动画对象。
	- parameter key:        动画对象的键。
	- parameter completion: 动画完成或被移除时调用此闭包，注意不要设置动画代理,否则会造成覆盖。
	*/
	func add(_ anim: CAAnimation, forKey key: String? = nil, completion: ((Bool) -> Void)?) {
		if let completion = completion {
			anim.delegate = AnimationDelegate(completion: completion)
		}
		base.add(anim, forKey: key)
	}

	/// 暂停、恢复动画。
	var paused: Bool {
		set {
			guard newValue != paused else { return }
			if newValue {
				/* speed 设置为 0.0 将导致图层本地时间变为 0.0，而无论 beginTime 和 timeOffset 之前的值是多少，
				将 timeOffset 设置为图层本地时间，就可以在 speed 设置为 0.0 后保持图层本地时间不变，
				此时 speed 为 0.0，动画停止，且图层本地时间未变，因此动画会停止在原处。*/
				base.timeOffset = base.convertTime(CACurrentMediaTime(), from: nil)
				base.speed = 0
			} else {
				/* speed 设置为 1.0 后，图层本地时间会恢复正常。此时若将 timeOffset 重置为 0.0，即 timeOffset = 0.0，
				图层本地时间将等于绝对时间 CACurrentMediaTime()。但是暂停的这段时间内图层本地时间并未增长，
				应该减去这段时间，这个时间增量刚好是 CACurrentMediaTime() - timeOffset。若 beginTime 不为 0.0，
				还应加上该值。因此最终结果为 timeOffset = 0 - (CACurrentMediaTime() - timeOffset) + beginTime。
				这样图层的本地时间就会等于暂停前的值，此时 speed 恢复为 1.0，且图层本地时间未变，因此动画会从原处恢复。*/
				base.speed = 1
				base.timeOffset += base.beginTime - CACurrentMediaTime()
			}
		}
		get {
			return base.speed == 0
		}
	}
}


